core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedValueResolver.java

```diff
@@ -1179,55 +1194,52 @@ public String toString() {
                               .add("context", context)
                               .add("request", request)
                               .add("aggregatedRequest", aggregatedRequest)
-                              .add("httpParameters", httpParameters)
+                              .add("queryParams", queryParams)
                               .toString();
         }
 
         /**
-         * Returns a map of parameters decoded from a request.
+         * Returns a {@link QueryParams} decoded from a request.
          *
          * <p>Usually one of a query string of a URI or URL-encoded form data is specified in the request.
          * If both of them exist though, they would be decoded and merged into a parameter map.</p>
          *
          * <p>Names and values of the parameters would be decoded as UTF-8 character set.</p>
          *
-         * @see QueryStringDecoder#QueryStringDecoder(String, boolean)
+         * @see QueryParams#fromQueryString(String)
          * @see HttpConstants#DEFAULT_CHARSET
          */
-        private static HttpParameters httpParametersOf(@Nullable String query,
-                                                       @Nullable MediaType contentType,
-                                                       @Nullable AggregatedHttpRequest message) {
+        private static QueryParams queryParamsOf(@Nullable String query,
+                                                 @Nullable MediaType contentType,
+                                                 @Nullable AggregatedHttpRequest message) {
             try {
-                Map<String, List<String>> parameters = null;
-                if (query != null) {
-                    parameters = new QueryStringDecoder(query, false).parameters();
-                }
-
+                final QueryParams params1 = query != null ? QueryParams.fromQueryString(query) : null;
+                QueryParams params2 = null;
                 if (message != null && isFormData(contentType)) {
                     // Respect 'charset' attribute of the 'content-type' header if it exists.
                     final String body = message.content(
                             contentType.charset().orElse(StandardCharsets.US_ASCII));
                     if (!body.isEmpty()) {
-                        final Map<String, List<String>> p =
-                                new QueryStringDecoder(body, false).parameters();
-                        if (parameters == null) {
-                            parameters = p;
-                        } else if (p != null) {
-                            parameters.putAll(p);
-                        }
+                        params2 = QueryParams.fromQueryString(body);
```

Comment:
Question: Just to make it sure that we can represent that form data as query parameters, right?

Reply:
Ah this reminds me I was thinking about this and forgot to comment. The name `QueryParams` might seem like it's tied to URL bar (I've had similar feeling about netty's names too). Perhaps UrlParams? It would be consistent with the mime type for form data, which is urlencoded.

Reply:
Hmm... isn't it just a query string sent as a body? `UrlParams` sounds more like it's part of a URL to me than the term 'query'.

Reply:
> Hmm... isn't it just a query string sent as a body? UrlParams sounds more like it's part of a URL to me than the term 'query'.

+1, I also think query params could be represented by either `GET` url or `POST` body.